modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java

```diff
@@ -0,0 +1,544 @@
+    }
+
+    /** */
+    private void createDumpLock() throws IgniteCheckedException, IOException {
+        File nodeDumpDir = IgniteSnapshotManager.nodeDumpDirectory(dumpDir, cctx);
+
+        if (!nodeDumpDir.mkdirs())
+            throw new IgniteCheckedException("Can't create node dump directory: " + nodeDumpDir.getAbsolutePath());
+
+        File lock = new File(nodeDumpDir, DUMP_LOCK);
+
+        if (!lock.createNewFile())
+            throw new IgniteCheckedException("Lock file can't be created or already exists: " + lock.getAbsolutePath());
+    }
+
+    /** */
+    private PartitionDumpContext dumpContext(int grp, int part) {
+        return dumpCtxs.computeIfAbsent(
+            toLong(grp, part),
+            key -> {
+                try {
+                    return new PartitionDumpContext(
+                        cctx.kernalContext().cache().cacheGroup(grp),
+                        part,
+                        new File(groupDirectory(cctx.cache().cacheGroup(grp)), PART_FILE_PREFIX + part + DUMP_FILE_EXT)
+                    );
+                }
+                catch (IOException | IgniteCheckedException e) {
+                    throw new IgniteException(e);
+                }
+            }
+        );
+    }
+
+    /**
+     * Context of dump single partition.
+     */
+    private class PartitionDumpContext implements Closeable {
+        /** Group id. */
+        final int grp;
+
+        /** Partition id. */
+        final int part;
+
+        /** Hashes of cache keys of entries changed by the user during partition dump. */
+        final Map<Integer, Set<KeyCacheObject>> changed;
+
+        /** Count of entries changed during dump creation. */
+        LongAdder changedCnt = new LongAdder();
+
+        /** Partition dump file. Lazily initialized to prevent creation files for empty partitions. */
+        final FileIO file;
+
+        /** Last version on time of dump start. Can be used only for primary. */
+        @Nullable final GridCacheVersion startVer;
+
+        /** Topology Version. */
+        private final AffinityTopologyVersion topVer;
+
+        /** Partition serializer. */
+        DumpEntrySerializer serdes;
+
+        /** If {@code true} context is closed. */
+        volatile boolean closed;
+
+        /** Count of writers. When count becomes {@code 0} context must be closed. */
+        private final AtomicInteger writers = new AtomicInteger(1); // Iterator writing entries to this context, by default.
+
+        /**
+         * @param gctx Group id.
+         * @param part Partition id.
+         * @param dumpFile Dump file path.
+         */
+        public PartitionDumpContext(CacheGroupContext gctx, int part, File dumpFile) throws IOException {
+            assert gctx != null;
+
+            this.part = part;
+            grp = gctx.groupId();
+            topVer = gctx.topology().lastTopologyChangeVersion();
+
+            boolean primary = grpPrimaries.computeIfAbsent(
```

Comment:
Let's calculate in CreateDumpFutureTask constructor in the `disco-notifier` thread. Instead of creating a bottle-neck during active `beforeChange` invokations.

Reply:
Fixed